app/models/rate_limit.rb

```diff
@@ -0,0 +1,99 @@
+require 'ipaddr'
+
+class RateLimit < ActiveRecord::Base
+  GLOB_PATTERN = /^(\*\*\.|\*\.)/
+  RECURSIVE_GLOB = "**."
+  RECURSIVE_PATTERN = "(?:[-a-z0-9]+\\.)+"
+  SINGLE_GLOB = "*."
+  SINGLE_PATTERN = "(?:[-a-z0-9]+\\.)"
+
+  validates :burst_rate, presence: true, numericality: { only_integer: true, greater_than: 0 }
+  validates :burst_period, presence: true, numericality: { only_integer: true, greater_than: 0 }
+  validates :sustained_rate, presence: true, numericality: { only_integer: true, greater_than: 0 }
+  validates :sustained_period, presence: true, numericality: { only_integer: true, greater_than: 0 }
+  validates :domain_whitelist, length: { maximum: 10000, allow_blank: true }
+  validates :ip_whitelist, length: { maximum: 10000, allow_blank: true }
+
+  validate do
+    unless sustained_rate.nil? || burst_rate.nil?
+      if sustained_rate <= burst_rate
+        errors.add :sustained_rate, "Sustained rate must be greater than burst rate"
+      end
+    end
+
+    unless sustained_period.nil? || burst_period.nil?
+      if sustained_period <= burst_period
+        errors.add :sustained_period, "Sustained period must be greater than burst period"
+      end
+    end
+
+    begin
+      domains
+    rescue StandardError => e
+      errors.add :domain_whitelist, :invalid
+    end
+
+    begin
+      ips
+    rescue StandardError => e
+      errors.add :ip_whitelist, :invalid
+    end
+  end
+
+  def exceeded?(signature)
+    return false if domain_whitelisted?(signature.domain)
+    return false if ip_whitelisted?(signature.ip_address)
+
+    burst_rate_exceeded?(signature) || sustained_rate_exceeded?(signature)
+  end
+
+  def domain_whitelist=(value)
+    @domains = nil
+    super(normalize_lines(value))
+  end
+
+  def ip_whitelist=(value)
+    @ips = nil
+    super(normalize_lines(value))
+  end
+
+  private
+
+  def domains
+    @domains ||= domain_whitelist.each_line.map{ |l| %r[\A#{convert_glob(l.strip)}\z] }
```

Comment:
Not a major issue, but it occurred to me that if the user enters blank lines it'll mean that they end up as '\A\z', which is a bit of an odd comparison.

Also, do you think it's worth allowing them to enter '#' as a comment? So they can structure/comment things?

```
# Some situation occurred date XYZ
example.org


# Some other situation occurred here
example.com
```